File: LeetCode/Study Plan/Practice Problems/Medium/Construct String With Repeat Limit/Go/construct_string.go
package main

import (
	"container/heap"
	"fmt"
)

type Pair struct {
	char byte
	cnt  int
}

type MaxHeap []Pair

func (p MaxHeap) Len() int              { return len(p) }
func (p MaxHeap) Less(i, j int) bool    { return p[i].char > p[j].char }
func (p MaxHeap) Swap(i, j int)         { p[i], p[j] = p[j], p[i] }
func (p *MaxHeap) Push(val interface{}) { *p = append(*p, val.(Pair)) }
func (p *MaxHeap) Pop() interface{} {
	val := (*p)[len(*p)-1]
	*p = (*p)[:len(*p)-1]
	return val
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func construct_string_with_repeat_limit(str string, repeat_limit int) string {
	count := make(map[byte]int)
	n := len(str)

	for i := 0; i < n; i++ {
		count[str[i]]++
	}

	mh := &MaxHeap{}
	for k, v := range count {
		heap.Push(mh, Pair{k, v})
	}

	s := ""
	for mh.Len() > 0 {
		curr := heap.Pop(mh).(Pair)
		c1 := curr.char
		cnt_c1 := curr.cnt

		t := min(repeat_limit, cnt_c1)
		for i := 0; i < t; i++ {
			s += string(c1)
		}

		if mh.Len() > 0 && cnt_c1-t > 0 {
			curr2 := heap.Pop(mh).(Pair)
			c2 := curr2.char
			cnt_c2 := curr2.cnt
			s += string(c2)
			if cnt_c1-t > 0 {
				heap.Push(mh, Pair{c1, cnt_c1 - t})
			}
			if cnt_c2-1 > 0 {
				heap.Push(mh, Pair{c2, cnt_c2 - 1})
			}
		}
	}
	return s
}

func main() {
	fmt.Println(construct_string_with_repeat_limit("cczazcc", 3))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Construct String With Repeat Limit/C++/construct_string.cpp
#include <iostream>
#include <cstdlib>
#include <algorithm>
#include <queue>
#include <string>
#include <map>
using namespace std;

string repeat_limited_string(string s, int repeat_limit) {
    string res = "";
    map<char,int> count;

    for (auto it : s){
        count[it]++;
    }

    priority_queue<pair<char,int> > pq;

    for (auto it : count){
        pq.push(make_pair(it.first,it.second));
    }

    while(!pq.empty()){
        char c1 = pq.top().first;
        int cnt_c1 = pq.top().second;
        pq.pop();

        int len = min(repeat_limit, cnt_c1);
        for (int i=0;i<len;i++){
            res+=c1;
        }

        if (!pq.empty() and cnt_c1-len>0){
            char c2 = pq.top().first;
            int cnt_c2 = pq.top().second;
            pq.pop();

            res+=c2;
            pq.push(make_pair(c1,cnt_c1-len));
            if (cnt_c2-1>0){
                pq.push(make_pair(c2, cnt_c2-1));
            }
        }
    }
    return res;
}


int main(){
    cout<<repeat_limited_string("cczazcc", 3)<<"\n";
    return 0;
}